utils/trigger_workflow.ts

```diff
@@ -0,0 +1,78 @@
+import axios from "axios"
+import { Runtime } from "wxt/browser"
+
+import { Workflow } from "@/utils/workflow"
+import { Event } from "@/utils/event"
+
+export const fetchWorkflow = async (tabURL: string): Promise<Workflow> => {
+  // TODO: Support enqueue of multiple workflows
+  const workflowIdsToEnqueue = new URL(tabURL).searchParams.getAll(
+    "fluidicWorkflowIds[]"
+  )
+  const [workflowId] = workflowIdsToEnqueue.map((s) => String(s).trim())
+  if (!workflowId) {
+    throw new Error(
+      "No workflowId found in tab url query params, url: " + tabURL
+    )
+  }
+
+  // Query workflowId as Workflow
+  // TODO: Replace with env-based api target OR propagate target in evt payload
+  const response = await axios.get(
+    `http://localhost:80/api/workflows/${workflowId}`
+  )
+  const workflow: Workflow = response.data
+  if (!workflow) {
+    throw new Error(`No workflow found for id ${workflowId} from response`)
+  }
+
+  return workflow
+}
+
+export const getOpenAIKey = (tabURL: string): string | null => {
+  const openAIKey = new URL(tabURL).searchParams.get("openAIKey")
+  return openAIKey
+}
+
+// Top-level background listener to init flows within content script inject components; must be in background.ts and async iife wrapped to avoid 'Error - user gesture required'
+export const topLevelBackgroundMsgListener = <M extends Message>(
+  message: M,
+  sender?: Runtime.MessageSender
+) => {
+  ;(async () => {
+    switch (message.event) {
+      case Event.TRIGGER_WORKFLOW_CHECK: {
+        const tab = sender?.tab
+        const { id: tabId, windowId, url: tabURL } = sender?.tab ?? {}
+        if (!tab || !tabId || !tabURL)
+          throw new Error( // TRIGGER_WORKFLOW_QUERY should not be sent until landing page has loaded, button has been injected, and tapped to open extension
+            `No tab id ${tabId} or tab url ${JSON.stringify(sender?.tab)} found from evt sender`
+          )
+
+        // Open side panel programatically — nesting or wrapping in any handler (without injecting a button from chrome runtime trigger by client) will lead to Error — user gesture required
+        // @ts-ignore
+        browser.sidePanel.open({ windowId })
+
+        // Query workflow data and begin internal execution of steps sequentially
+        const workflow = await fetchWorkflow(tabURL)
+        const tabs = await browser.tabs.query({
+          active: true,
+          currentWindow: true,
+        })
+        const anchorHeadlessTab = tabs.find((t) =>
+          t.url?.includes("headless/home?fluidicWorkflowIds[]")
```

Comment:
I disagree with this actually @jrpotter but we should def chat through it... IMO anything internal and/or involving user/internal-specific data should be protected within claims in JWT on the user's physical browser client. We can append keys on the backend as needed (as not to expose an open ai key) since we can enrich the original button link on the backend before exec. So JWTs should be sent (or local client state) for claims to access anything sensitive within the workflow. Workflow ids, as well as any others ids that are long-lived for that matter, should always be accessed (if not public) within the scope of a session/JWT.

Reply:
That session/JWT could be anonymous with minimal claims btw